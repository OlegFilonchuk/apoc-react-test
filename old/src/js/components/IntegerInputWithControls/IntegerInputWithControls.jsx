import React from 'react';
import PropTypes from 'prop-types';
import { isFinite } from 'lodash';
import classnames from 'classnames';

import Input from '../Input/Input';
import Label from '../Label/Label';
import Button from '../Button/Button';

import './IntegerInputWithControls.less';

class IntegerInputWithControls extends React.Component {
  static checkIfZero = val => Number(val) === 0;

  static isValidPositiveInteger = val => val % 1 === 0 && isFinite(Number(val)) && Number(val) >= 0;

  constructor(props) {
    super(props);

    this.state = {
      isZero: IntegerInputWithControls.checkIfZero(props.value),
      value: props.value
    };
  }

  onChange = e => {
    const {
      target: { value }
    } = e;

    if (IntegerInputWithControls.isValidPositiveInteger(value)) {
      this.setState({
        isZero: IntegerInputWithControls.checkIfZero(value),
        value
      });

      this.props.onValueChange({ key: this.props.label, value });
    }
  };

  get wrapperClassName() {
    return classnames('integer-input-with-controls', this.props.wrapperClassName);
  }

  get labelText() {
    return this.props.children && this.props.children.length ? this.props.children : this.props.label;
  }

  get inputProps() {
    return { ...this.props, children: null };
  }

  addition = () => {
    this.genericButtonClick(Number(this.state.value) + 1);
  };

  subtraction = () => {
    this.genericButtonClick(Number(this.state.value) - 1);
  };

  genericButtonClick = value => {
    this.setState({
      isZero: IntegerInputWithControls.checkIfZero(value),
      value: `${value}`
    });

    this.props.onValueChange({ key: this.props.label, value });
  };

  render() {
    const { value, isZero } = this.state;

    return (
      <div className={this.wrapperClassName}>
        <Label htmlFor={this.props.label}>{this.labelText}</Label>
        <Button className="sc-btn-primary-outline icon-button" onClick={this.addition}>
          <span className="sc-icon-plus" />
        </Button>
        <Button disabled={isZero} className="sc-btn-primary-outline icon-button" onClick={this.subtraction}>
          <span className="sc-icon-solid-square" />
        </Button>
        <Input {...this.inputProps} id={this.props.label} onChange={this.onChange} value={value} />
      </div>
    );
  }
}

IntegerInputWithControls.propTypes = {
  /**
   * Any custom class names for wrapper
   */
  wrapperClassName: PropTypes.string,

  /**
   * Input value
   */
  value: PropTypes.string,

  /**
   * Label value that will also be used as id for base input
   */
  label: PropTypes.string,

  /**
   * onValueChange should be passed to this component
   */
  onValueChange: PropTypes.func,

  /**
   * children can be passed to act like a label of input
   */
  children: PropTypes.node
};

IntegerInputWithControls.defaultProps = {
  wrapperClassName: '',
  value: '0',
  label: 'Label',
  children: null,
  onValueChange: () => {}
};

export default IntegerInputWithControls;
