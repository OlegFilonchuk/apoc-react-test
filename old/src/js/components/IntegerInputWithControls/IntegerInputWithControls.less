.integer-input-with-controls {
  display: flex;
  justify-content: flex-end;
  align-items: center;

  label {
    margin: 0;
    flex: 1 1 auto;
  }

  button {
    height: 16px;
    width: 16px;
    padding: 0;
    margin: 0 5px 0 0;

    span.sc-icon-solid-square,
    span.sc-icon-plus {
      font-size: 8px;
    }

    span.sc-icon-solid-square {
      transform: rotateX(60deg) translate(-50%, -90%) scaleX(1.5);
    }
  }

  input {
    width: 40px;
    height: 20px;
  }
}
