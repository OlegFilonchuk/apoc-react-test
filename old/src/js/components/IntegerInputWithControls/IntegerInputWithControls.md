Allow to render [`<Input>`](#input) together with two [`<Button>`](#button) components, designated to increase/decrease integer value inside [`<Input>`](#input) by 1.

You can pass all of the [`<Input>` props](#input) to the `<IntegerInputWithControls>` component.
Please pass `wrapperClassName` prop to adjust `<IntegerInputWithControls>` styles.
Passed `label` prop will be used as label text and as `id` of base input.

# Example
    (
      <IntegerInputWithControls
        label="Some label"
        value="2"
      />
    )
