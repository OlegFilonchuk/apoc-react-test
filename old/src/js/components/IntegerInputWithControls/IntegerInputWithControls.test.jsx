import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import IntegerInputWithControls from './IntegerInputWithControls';

describe('<IntegerInputWithControls />', () => {
  const renderInput = (props = {}, mountFn = shallow) => mountFn(<IntegerInputWithControls {...props} />);

  describe('in snapshot tests', () => {
    it('should render correctly without props', () => {
      const input = renderInput();

      expect(toJson(input)).toMatchSnapshot();
    });

    it('should render Input with custom className', () => {
      const input = renderInput({ wrapperClassName: 'some-class-name' });

      expect(toJson(input)).toMatchSnapshot();
    });

    it('should render Input with custom value', () => {
      const input = renderInput({ value: '1' });

      expect(toJson(input)).toMatchSnapshot();
    });

    it('should render label with value passed as child node', () => {
      const input = shallow(<IntegerInputWithControls>Very fancy label</IntegerInputWithControls>);

      expect(toJson(input)).toMatchSnapshot();
    });
  });

  describe('in logic tests', () => {
    it('should render with disabled minus button but enabled plus button by default', () => {
      const input = renderInput();

      expect(input.state().isZero).toBeTruthy();
    });

    it('should correctly handle normal addition and subtraction via buttons', () => {
      const input = renderInput({ value: '1' }, mount);

      expect(input.state().isZero).toBeFalsy();
      input
        .find('button')
        .last()
        .simulate('click');
      expect(input.state().isZero).toBeTruthy();
      input
        .find('button')
        .first()
        .simulate('click');
      expect(input.state().isZero).toBeFalsy();
    });

    it('should correctly handle manual input of different values', () => {
      const input = renderInput({}, mount);

      expect(input.state().isZero).toBeTruthy();

      input.find('input').simulate('change', { target: { value: '123' } });
      expect(input.state().isZero).toBeFalsy();

      input.find('input').simulate('change', { target: { value: '' } });
      expect(input.state().isZero).toBeTruthy();

      input.find('input').simulate('change', { target: { value: 'adsf' } });
      expect(input.state().isZero).toBeTruthy();
    });
  });
});
