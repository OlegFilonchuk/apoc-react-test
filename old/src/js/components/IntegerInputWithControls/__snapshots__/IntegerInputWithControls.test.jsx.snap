// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IntegerInputWithControls /> in snapshot tests should render Input with custom className 1`] = `
<div
  className="integer-input-with-controls some-class-name"
>
  <Label
    className=""
    htmlFor="Label"
    required={false}
  >
    Label
  </Label>
  <Button
    className="sc-btn-primary-outline icon-button"
    data-test-element=""
    disabled={false}
    id=""
    isLinkView={false}
    isSelected={false}
    name=""
    onClick={[Function]}
    title=""
    type="button"
    value={null}
  >
    <span
      className="sc-icon-plus"
    />
  </Button>
  <Button
    className="sc-btn-primary-outline icon-button"
    data-test-element=""
    disabled={true}
    id=""
    isLinkView={false}
    isSelected={false}
    name=""
    onClick={[Function]}
    title=""
    type="button"
    value={null}
  >
    <span
      className="sc-icon-solid-square"
    />
  </Button>
  <Input
    className=""
    hasError={false}
    id="Label"
    label="Label"
    lettersLeftCount={Infinity}
    onChange={[Function]}
    onInputRef={[Function]}
    onKeyDown={[Function]}
    onValueChange={[Function]}
    submitOnEnter={false}
    suffix=""
    type="text"
    value="0"
    wrapperClassName="some-class-name"
  />
</div>
`;

exports[`<IntegerInputWithControls /> in snapshot tests should render Input with custom value 1`] = `
<div
  className="integer-input-with-controls"
>
  <Label
    className=""
    htmlFor="Label"
    required={false}
  >
    Label
  </Label>
  <Button
    className="sc-btn-primary-outline icon-button"
    data-test-element=""
    disabled={false}
    id=""
    isLinkView={false}
    isSelected={false}
    name=""
    onClick={[Function]}
    title=""
    type="button"
    value={null}
  >
    <span
      className="sc-icon-plus"
    />
  </Button>
  <Button
    className="sc-btn-primary-outline icon-button"
    data-test-element=""
    disabled={false}
    id=""
    isLinkView={false}
    isSelected={false}
    name=""
    onClick={[Function]}
    title=""
    type="button"
    value={null}
  >
    <span
      className="sc-icon-solid-square"
    />
  </Button>
  <Input
    className=""
    hasError={false}
    id="Label"
    label="Label"
    lettersLeftCount={Infinity}
    onChange={[Function]}
    onInputRef={[Function]}
    onKeyDown={[Function]}
    onValueChange={[Function]}
    submitOnEnter={false}
    suffix=""
    type="text"
    value="1"
    wrapperClassName=""
  />
</div>
`;

exports[`<IntegerInputWithControls /> in snapshot tests should render correctly without props 1`] = `
<div
  className="integer-input-with-controls"
>
  <Label
    className=""
    htmlFor="Label"
    required={false}
  >
    Label
  </Label>
  <Button
    className="sc-btn-primary-outline icon-button"
    data-test-element=""
    disabled={false}
    id=""
    isLinkView={false}
    isSelected={false}
    name=""
    onClick={[Function]}
    title=""
    type="button"
    value={null}
  >
    <span
      className="sc-icon-plus"
    />
  </Button>
  <Button
    className="sc-btn-primary-outline icon-button"
    data-test-element=""
    disabled={true}
    id=""
    isLinkView={false}
    isSelected={false}
    name=""
    onClick={[Function]}
    title=""
    type="button"
    value={null}
  >
    <span
      className="sc-icon-solid-square"
    />
  </Button>
  <Input
    className=""
    hasError={false}
    id="Label"
    label="Label"
    lettersLeftCount={Infinity}
    onChange={[Function]}
    onInputRef={[Function]}
    onKeyDown={[Function]}
    onValueChange={[Function]}
    submitOnEnter={false}
    suffix=""
    type="text"
    value="0"
    wrapperClassName=""
  />
</div>
`;

exports[`<IntegerInputWithControls /> in snapshot tests should render label with value passed as child node 1`] = `
<div
  className="integer-input-with-controls"
>
  <Label
    className=""
    htmlFor="Label"
    required={false}
  >
    Very fancy label
  </Label>
  <Button
    className="sc-btn-primary-outline icon-button"
    data-test-element=""
    disabled={false}
    id=""
    isLinkView={false}
    isSelected={false}
    name=""
    onClick={[Function]}
    title=""
    type="button"
    value={null}
  >
    <span
      className="sc-icon-plus"
    />
  </Button>
  <Button
    className="sc-btn-primary-outline icon-button"
    data-test-element=""
    disabled={true}
    id=""
    isLinkView={false}
    isSelected={false}
    name=""
    onClick={[Function]}
    title=""
    type="button"
    value={null}
  >
    <span
      className="sc-icon-solid-square"
    />
  </Button>
  <Input
    className=""
    hasError={false}
    id="Label"
    label="Label"
    lettersLeftCount={Infinity}
    onChange={[Function]}
    onInputRef={[Function]}
    onKeyDown={[Function]}
    onValueChange={[Function]}
    submitOnEnter={false}
    suffix=""
    type="text"
    value="0"
    wrapperClassName=""
  />
</div>
`;
